### PR TITLE
Run preload at import, start, and teardown

### PR DESCRIPTION
**Pull request by mrocklin**
Previously we only ran preload scripts after the server had started.
This made it difficult to modify the server before certain actions had
taken place.  For example, a test in this PR registers a new Comm
backend for the server to use.  This would not have been possible
before.

Now we run different preload functions when

1.  we first instantiate the server
2.  we first start the server
3.  we close the server (we used to run teardown at `atexit` time

cc @pentschev I think that you asked for this at one point.

### CODE REVIEW

**Comment by mrocklin** [comment]
Merging in 24 hours if there are no comments

File: distributed/preloading.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def _import_module(name, file_dir=None):
     }
 
 
-def preload_modules(names, parameter=None, file_dir=None, argv=None):
+def on_creation(names, file_dir=None):
```
Comment (jrbourbeau) [inline distributed/preloading.py:114]: We can remove `parameter` and `argv` from the docstring for this function

File: distributed/preloading.py [hunk 2]
```diff
@@ -128,11 +128,25 @@ def preload_modules(names, parameter=None, file_dir=None, argv=None):
     if isinstance(names, str):
         names = [names]
 
-    for name in names:
-        interface = _import_module(name, file_dir=file_dir)
+    return {name: _import_module(name, file_dir=file_dir) for name in names}
 
+
+async def on_start(names, parameter=None, file_dir=None, argv=None):
+    """ Imports modules, handles `dask_setup` and `dask_teardown`.
```
Comment (jrbourbeau) [inline distributed/preloading.py:135]:
```suggestion
    """ Imports modules and handles ``dask_setup``.
```

File: distributed/preloading.py [hunk 3]
```diff
@@ -128,11 +128,25 @@ def preload_modules(names, parameter=None, file_dir=None, argv=None):
...
+
+    Parameters
+    ----------
+    names: list of strings
+        Module names or file paths
+    parameter: object
+        Parameter passed to `dask_setup` and `dask_teardown`
```
Comment (jrbourbeau) [inline distributed/preloading.py:142]:
```suggestion
        Parameter passed to ``dask_setup``
```

File: distributed/preloading.py [hunk 4]
```diff
@@ -141,8 +155,25 @@ def preload_modules(names, parameter=None, file_dir=None, argv=None):
...
             else:
-                dask_setup(parameter)
+                future = dask_setup(parameter)
+                if inspect.isawaitable(future):
+                    await future
                 logger.info("Run preload setup function: %s", name)
 
+
+async def on_teardown(names, parameter=None, file_dir=None):
+    """ Imports modules, handles `dask_setup` and `dask_teardown`.
```
Comment (jrbourbeau) [inline distributed/preloading.py:165]:
```suggestion
    """ Imports modules and handles ``dask_teardown``.
```

File: distributed/preloading.py [hunk 5]
```diff
@@ -141,8 +155,25 @@ def preload_modules(names, parameter=None, file_dir=None, argv=None):
...
+
+    Parameters
+    ----------
+    names: list of strings
+        Module names or file paths
+    parameter: object
+        Parameter passed to `dask_setup` and `dask_teardown`
```
Comment (jrbourbeau) [inline distributed/preloading.py:172]:
```suggestion
        Parameter passed to ``dask_teardown``
```

File: distributed/preloading.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import atexit
```
Comment (jrbourbeau) [inline distributed/preloading.py:1]: +1 I like that we're calling the teardown code explicitly in in `close` methods now instead of using `atexit`

Comment (mrocklin) [reply distributed/preloading.py:1]: Yeah, I figure that if people want to run atexit they can use `atexit` in their preload module itself.  We don't have to mandate that structure.

**Comment by mrocklin** [comment]
Thanks for the review @jrbourbeau .  I believe I've handled everything.  I also changed things up slightly so that we're now passing around the imported modules.

I plan to follow up on this PR with a PR that supports preloads for the Nanny.

**Pull request merged**